### DISCUSSION TITLE
Anisotropic material

### DISCUSSION THREAD

**Discussion by rugfri** [PyEllips/pyElli#149] [Q&A]
Dear PyElli's,

I would like to build a model using the UniaxialMaterial class and I am not sure about a couple of things, hope you can help. I did not find  open/closed issues/discussions. In case this is a duplicate please point me there.

I am going along these lines, to build a stack where I report only the meaningful (the one I believe is) part of the code:
For the case of a thin film on SiO2 on Si

```python
@fit(psi_delta, params)
def model(lbda, params):
    ## dispersion ordinary axis
    disp_o = elli.LorentzEnergy().add(
        params["amplitude_o"],
        params["energy_o"],
        params["broadening_o"],)
    ## dispersion extra-ordinary axis
    disp_e = elli.LorentzEnergy().add(
        params["amplitude_e"],
        params["energy_e"],
        params["broadening_e"],)
    epi = elli.materials.UniaxialMaterial(disp_o, disp_e)
    
    SiO2 = elli.Cauchy(
        params["SiO2_n0"],
        params["SiO2_n1"],).get_mat()

    stack = [elli.Layer(epi, params["thickness"]), elli.Layer(SiO2, params["SiO2_t"])]
    return elli.Structure(elli.AIR, stack, Si).evaluate(lbda, angle, solver=elli.Solver4x4).as_delta_range(0,180)
```
 
How the anisotropic layer is assumed to be oriented? The extraordinary axis is normal to the substrate surface?
How would one here add an EpsilonInf term to the uniaxial layer?
I am using the 4x4 solver, correct? 
The  "as_delta_range(0,180)" is due to Woollam data.

Thanks,
Best regards
Ruggero

**Comment by domna** [comment]
Hey Ruggero,

I don't remember a previous issue on this as well. Your code looks very good already and you just have to do a few tweaks to account for correct alignment of the sample.



> How the anisotropic layer is assumed to be oriented? The extraordinary axis is normal to the substrate surface?

There is no general relation of the extraordinary axis to, e.g., the sample surface. This is because you could have a different cut of your material which would mean that your sample geometry has another relation to ordinary/extraordinary axes. For an experiment you either know your alignment of the optical axes to the sample surfaces already or you have to experimentally determine this. You can also fit this in pyElli if you want! However, generally you should at least roughly know in with direction your optical axes w.r.t. to the geometry are situated. Otherwise the parameter space of 3D rotation is just to large.

Now, how do you align the sample geometry to the optical axes? In pyElli we do this by rotating the material with a suitable rotation matrix. You can see this in the [Cholesteric Liquid Crystal example](https://pyelli.readthedocs.io/en/stable/auto_examples/plot_cholesteric_lq.html). This is the most relevant part for you:

```python
LC = elli.UniaxialMaterial(
    elli.ConstantRefractiveIndex(no), elli.ConstantRefractiveIndex(ne)
)  # ne is along z
R = elli.rotation_v_theta(elli.E_Y, 90)  # rotation of pi/2 along y
LC.set_rotation(R)  # apply rotation from z to x
```
As you see you can just apply a rotation matrix by calling  `.set_rotation` on your material. [Here](https://pyelli.readthedocs.io/en/stable/helpers.html#elli.utils.rotation_euler) you see a list of helper functions for creating rotation matrices.

> How would one here add an EpsilonInf term to the uniaxial layer?

The `EpsilonInf` term is not added to the material (i.e. layer) but the dispersions. So you either add two different `EpsilonInf` to the `disp_` variables or just use the same fitting parameter if you wish to use the same (but generally I would say that these are probably different).
This could look like this:

```python
disp_e = elli.LorentzEnergy().add(
        params["amplitude_e"],
        params["energy_e"],
        params["broadening_e"],) + params["epsilon_inf"]
```

A number just adds as epsiloninf (i.e. its the same as using `EpsilonInf(params["epsilon_inf"])`). This should work, however, I typed it from the top of my head and did not test it.

> I am using the 4x4 solver, correct?

Indeed, the 2x2 cannot account for anisotropy as it treats s and p waves independently and w/o cross-interaction.

> The "as_delta_range(0,180)" is due to Woollam data.

👍️

**Comment by MarJMue** [comment]
> How the anisotropic layer is assumed to be oriented? The extraordinary axis is normal to the substrate surface?

Without rotation, the Materiel is oriented with the extraordinary axis perpendicular to the layer surface. 

If you prefer the euler angles, in the Fujiwara's book _Spectroscopic Ellipsometry - Principles and Applications_ Figure 6.10 has some common examples.

**Reply by rugfri** [reply]
Hi both, thanks.

So, as defined
epi = elli.materials.UniaxialMaterial(disp_o, disp_e)
has the axes are oriented as in fig. 6.10a, the rotation explained by Florian brings the axes as in fig. 6.10b. Correct?
..and using the various elli.utils one 

A bit more speculative, how would be best to approach the case of a domain-like structure of an uniaxial material with the unique axis is along x (or y)?
Of course things can be complex, e.g. domains size in relation to the beam size..
Let's say the beam size is larger than the domain size, so that the experiment provides an averaging. 
I guess one option is simply assuming the material isotropic ..
Else, some sort of effective medium/mixture?

Thanks!

**Reply by MarJMue** [reply]
> So, as defined epi = elli.materials.UniaxialMaterial(disp_o, disp_e) has the axes are oriented as in fig. 6.10a, the rotation explained by Florian brings the axes as in fig. 6.10b. Correct? ..and using the various elli.utils one

Florians example rotates around the y-axis. `elli.utils.rotation_euler(0,90,0)` (using the same angles as in the book) is equivalent to `elli.rotation_v_theta(elli.E_X, 90)`.

> A bit more speculative, how would be best to approach the case of a domain-like structure of an uniaxial material with the unique axis is along x (or y)? [...] Let's say the beam size is larger than the domain size, so that the experiment provides an averaging.

Assuming you have random distribution of your unique axis, i think a isotropic model with averaged parameters would be the best option. There might be a case for averaging the result for simulations with all possible orientations, but thats something i would need to test myself.

If there is an preferred direction you might be able to fit the angle. But i haven't work with material with large domains at all and can only speculate.

I have successfully used pyelli for the case, that the domains are bigger than the beam size (single crystals as extreme case) and it works really good.

**Reply by rugfri** [reply]
Thanks.

I have a case where the unique axis is not randomly distributed in 3D, but only in 2D.
For example the unique axis of the anisotropic layer is parallel to the substrate surface but does not have a preferred direction within the plane.
What would be the best modelling approach for such a case?

Thanks!

**Reply by MarJMue** [reply]
Sorry, I did not answer quicker.
You could either use a Model with an averaged dispersion for the parallel axis and another for the perpendicular axis.

Or you could use a new feature, which I have not shown in any example. You could use the physically correct material and generate multiple structures, while you rotate the material accordingly. Then you could use the [ResultList](https://pyelli.readthedocs.io/en/stable/result.html#elli.result.ResultList) to average over all possible rotations.

**Reply by rugfri** [reply]
Thanks, I will the new feature a try. Sounds interesting!
